File: frontend/src/types/status.ts
import type { EntityConfig } from './domain';

export type SpecimenState = 'received' | 'testing' | 'hold' | 'released' | 'disposed';
export const ALL_SPECIMEN_STATE: readonly SpecimenState[] = ['received', 'testing', 'hold', 'released', 'disposed'];
export type SignoffState = 'draft' | 'peer_review' | 'signed' | 'rejected';
export const ALL_SIGNOFF_STATE: readonly SignoffState[] = ['draft', 'peer_review', 'signed', 'rejected'];

export const ENTITY_CONFIGS: readonly EntityConfig[] = [
  { key: 'animalCase', path: 'cases', label: '动物样本来源', statuses: ['registered', 'sampling', 'testing', 'closed'] as const, primaryTransitions: { registered: 'sampling', sampling: 'testing', testing: 'closed' } },
  { key: 'specimen', path: 'specimens', label: '检验样本', statuses: ['received', 'testing', 'hold', 'released', 'disposed'] as const, primaryTransitions: { received: 'testing', testing: 'released', hold: 'released', released: 'disposed' } },
  { key: 'assayRun', path: 'assays', label: '检测运行', statuses: ['planned', 'running', 'validated', 'invalid'] as const, primaryTransitions: { planned: 'running', running: 'validated' } },
  { key: 'resultSignoff', path: 'signoff', label: '结果签发', statuses: ['draft', 'peer_review', 'signed', 'rejected'] as const, primaryTransitions: { draft: 'peer_review', peer_review: 'signed' } }
];

File: backend/internal/constants/status.go
package constants

// Shared status values are mirrored in frontend/src/types/status.ts. Keeping
// the lists explicit makes state-machine drift visible during code review.

type SpecimenState string

const (
	SpecimenStateReceived SpecimenState = "received"
	SpecimenStateTesting  SpecimenState = "testing"
	SpecimenStateHold     SpecimenState = "hold"
	SpecimenStateReleased SpecimenState = "released"
	SpecimenStateDisposed SpecimenState = "disposed"
)

var AllSpecimenState = []string{"received", "testing", "hold", "released", "disposed"}

type SignoffState string

const (
	SignoffStateDraft      SignoffState = "draft"
	SignoffStatePeerReview SignoffState = "peer_review"
	SignoffStateSigned     SignoffState = "signed"
	SignoffStateRejected   SignoffState = "rejected"
)

var AllSignoffState = []string{"draft", "peer_review", "signed", "rejected"}

var AnimalCaseTransitions = map[string]map[string]bool{
	"registered": {"sampling": true, "testing": true},
	"sampling":   {"testing": true, "closed": true, "registered": true},
	"testing":    {"closed": true, "sampling": true},
	"closed":     {"testing": true},
}

var SpecimenTransitions = map[string]map[string]bool{
	"received": {"testing": true, "hold": true},
	"testing":  {"hold": true, "released": true, "received": true},
	"hold":     {"released": true, "disposed": true, "testing": true},
	"released": {"disposed": true, "hold": true},
	"disposed": {"released": true},
}

var AssayRunTransitions = map[string]map[string]bool{
	"planned":   {"running": true, "validated": true},
	"running":   {"validated": true, "invalid": true, "planned": true},
	"validated": {"invalid": true, "running": true},
	"invalid":   {"validated": true},
}

var ResultSignoffTransitions = map[string]map[string]bool{
	"draft":       {"peer_review": true},
	"peer_review": {"signed": true, "rejected": true},
	"signed":      {},
	"rejected":    {},
}

func CanTransition(graph map[string]map[string]bool, from, to string) bool {
	targets, exists := graph[from]
	return exists && targets[to]
}
